datalad_next/itertools/reroute.py

```diff
@@ -0,0 +1,232 @@
+""" Functions that allow to route data around upstream iterator """
+
+from __future__ import annotations
+
+from typing import (
+    Any,
+    Callable,
+    Generator,
+    Iterable,
+)
+
+
+__all__ = ['dont_process', 'join_with_list', 'route_in', 'route_out']
+
+
+dont_process = object()
```

Comment:
I will move to the pattern of using custom types.
I agree that the name should be improved. I think `DontYield`, `Drop`, `Ignore`, or `Nothing` would fit, since they appear in the element of the result-tuple that is either yielded or not yielded. Looking at them:

Some examples in which `data.name` is stored and nothing should be yielded to the consumer:

```python
return (DontYield, data.name)

return (Drop, data.name)

return (Ignore, data.name)

return (Nothing, data.name)

```

Alternatively we could use a dataclass for the result of a splitter-function which would make the semantics clearer:

```python

return SplitterResult(yield_value=DontYield, store=data.name)

return SplitterResult(yield_value=Drop, store=data.name)

return SplitterResult(yield_value=Ignore, store=data.name)

return SplitterResult(yield_value=Nothing, store=data.name)
```